include/hipSYCL/runtime/instrumentation.hpp

```diff
@@ -0,0 +1,153 @@
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef HIPSYCL_INSTRUMENTATION_HPP
+#define HIPSYCL_INSTRUMENTATION_HPP
+
+#include <cstdint>
+#include <chrono>
+#include <future>
+#include <unordered_map>
+#include <typeindex>
+#include <type_traits>
+
+namespace hipsycl {
+namespace rt {
+
+// SYCL defines timestamps as uint64_t with nanosecond resolution
+class profiler_clock {
+public:
+  using rep = uint64_t;
+  using period = std::nano;
+  using duration = std::chrono::duration<rep, period>;
+  using time_point = std::chrono::time_point<profiler_clock>;
+
+  constexpr static bool is_steady = true;
+
+  static time_point now() {
+    return time_point{duration{std::chrono::steady_clock::now().time_since_epoch()}};
+  }
+};
+
+class instrumentation {
+public:
+  virtual ~instrumentation() = default;
+};
+
+template<typename T>
+inline constexpr bool is_instrumentation_v
+    = std::is_convertible_v<std::remove_cv_t<T> *, instrumentation *>;
+
+class not_instrumented: public std::exception {
+public:
+  explicit not_instrumented(const std::type_info &instr_type): _instr_type(&instr_type) {}
+  const char *what() const noexcept override { return _instr_type->name(); }
+  const std::type_info &instrumentation() const { return *_instr_type; }
+
+private:
+  const std::type_info *_instr_type;
+};
+
+// Provides synchronization points for instrumentations between front- and backend threads.
+//
+// Thread safety: Before submission - single-threaded access only
+//                After submission: - concurrent access, but do not call instrument<>()
+class instrumentation_set {
+public:
+  // Create a producer-consumer channel for a particular instrumentation
+  template<typename Instr>
+  void instrument() {
```

Comment:
I wonder if it might be better to have a single class for instrumentation (as in my original proposal) instead of distinct classes. This way it would be easier to share state between them, which I believe is going to be a common case as often times one instrumentation mechanism is going to yield multiple values.

I'm not yet entirely convinced that the added complexity of this design is worth the additional capability of being able to return non-double values (which might never actually be needed in practice).

Reply:
Hmm, difficult to say for me without concrete examples of an API for other, future instrumentations. I suspect that restricting profiling info to any single type will turn out to be too restrictive.

`timestamp_profiler` is already an example how state sharing can be achieved - it yields 3 values and cannot be subdivided further. The requirement my design introduces, though, is that the same amount of state-sharing is appropriate between all backend implementations of an instrumentation.

In terms of complexity, I don't think there is a lot to be gained. The one thing we can remove by restricting return types are the template parameters on the methods of `instrumentation_set` (which would then be called `instrumentation`), and we would also be able to merge `instrumentation_set::await()` with `timestamp_profiler::await()`. The channel logic is still required.